reverseproxy/keyval.go

```diff
@@ -59,3 +62,46 @@ func removeVoucher(col keyValCollection, prefix string) {
 	col.Del(prefix + AMOUNT_VOUCHER_PARAM)
 	col.Del(prefix + SIGNATURE_VOUCHER_PARAM)
 }
+
+const (
+	RANGE_HEADER = "Range"
+	UNIT_PREFIX  = "bytes="
+)
+
+// hasRangeHeader returns true if the given header has a range header value.
+func hasRangeHeader(h http.Header) bool {
+	return h.Get(RANGE_HEADER) != ""
+}
+
+// parseRangeHeader parses the range header and returns the total number of bytes requested.
+func parseRangeHeader(h http.Header) (total uint64, err error) {
+	if !hasRangeHeader(h) {
+		return 0, fmt.Errorf("no range header")
+	}
+	rangeVal := h.Get(RANGE_HEADER)
+	var hasPrefix bool
+	rangeVal, hasPrefix = strings.CutPrefix(rangeVal, UNIT_PREFIX)
+	if !hasPrefix {
+		return 0, fmt.Errorf("range header value '%s' missing prefix 'byte='", rangeVal)
+	}
+
+	// It is valid to have multiple ranges, separated by commas. IE: bytes=0-499,1000-1499
+	for _, r := range strings.Split(rangeVal, ",") {
+		// Each range is a start and end separated by a dash. IE: 0-499
+		numVals := strings.Split(r, "-")
+
+		start, err := strconv.ParseUint(numVals[0], 10, 64)
+		if err != nil {
+			return 0, fmt.Errorf("could not parse start: %w", err)
+		}
+
+		end, err := strconv.ParseUint(numVals[1], 10, 64)
+		if err != nil {
+			return 0, fmt.Errorf("could not parse start: %w", err)
+		}
+
```

Comment:
We might want to validate that `end > start`. Otherwise, could the following scenario arise:
- the caller requests a valid, positive range and an invalid negative range.
- the destination server services the valid range.
- the caller underpays for the valid range.

Reply:
https://github.com/statechannels/go-nitro/pull/1493/commits/68f81a64d8f36670af6f8fd95ee7b0b618a170bf